{{cookiecutter.project_slug}}/backend/app/app/crud/base.py

```diff
@@ -0,0 +1,161 @@
+from typing import List, Optional
+
+from fastapi.encoders import jsonable_encoder
+from sqlalchemy.orm import Session
+
+from app.db.base_class import Base
+from pydantic import BaseModel
+
+
+class CrudBase:
+
+    def __init__(self, db_model: Base):
+        """
+        CrudBase instances are used to provide the basic CRUD methods for a given object type (get, get_multi, update, create and delete).
+        
+        In order to use it, follow this steps when you define a new DB model:
+        - create a class that inherites from CrudBase
+        - override basic methods with proper types (to get better completion in your IDE)
+        - create an instance of your newly created class, providing the DB model as an argument
+        
+        E.g.:
+
+            # model definition in app/models/item.py
+            class ItemCreate(...)
+                ...
+
+            class ItemUpdate(...)
+                ...
+
+            # model definition in app/db_models/item.py
+            class Item(Base):
+                id: int
+                ...
+
+            # crud definition in app/crud/item.py
+            from app.db_models.item import Item
+            from app.models.item import ItemUpdate, ItemCreate
+            from app.crud.base import CrudBase
+
+
+            class CrudItem(CrudBase):
+
+                def get(self, db_session: Session, obj_id: int) -> Optional[Item]:
+                    return super(CrudItem, self).get(db_session, obj_id=obj_id)
+
+                def get_multi(self, db_session: Session, *, skip=0, limit=100) -> List[Optional[Item]]:
+                    return super(CrudItem, self).get_multi(db_session, skip=skip, limit=limit)
+
+                def create(self, db_session: Session, *, obj_in: ItemCreate) -> Item:
+                    return super(CrudItem, self).create(db_session, obj_in=obj_in)
+
+                def update(self, db_session: Session, *, obj: Base, obj_in: ItemUpdate) -> Item:
+                    return super(CrudItem, self).update(db_session, obj=obj, obj_in=obj_in)
+
+
+            crud_item = CrudItem(Item)
+
+        Arguments:
+            db_model {Base} -- Class of the DB model which CRUD methods will be provided for
+        """  # noqa
+        self.db_model = db_model
+
+    def get(self, db_session: Session, obj_id: int) -> Optional[Base]:
```

Comment:
Here's how to add a get_by method that will let you pass in arbitraty attributes:

```py
    def get_by(self, db_session: Session, **kwargs: Any) -> Optional[Base]:
        filter_kwargs = {getattr(self.db_model, k)==v for k, v in kwargs.items()}
        return db_session.query(self.db_model).filter(*filter_kwargs).first()
```

Reply:
The idea looks good but I am afraid I am not up to the latest trend: I don't understand what `filter_kwargs`is ...

it looks like a set of booleans... but that does not make sense does it ?

Could you explain or give me a few pointers to the doc please 😃

Reply:
SQLAchemy uses statements like `Model.field==value` to generate an object, that can be passed as an argument to perform filtering. So, it's a bit misleading but when you see:

`db_session.query(db_model).filter(db_model.field==value).first()`

It's the same as 
```
filter_args=(db_model.field==value)
db_session.query(db_model).filter(filter_args).first()
```

```
filter_args=(getattr(db_model, 'field')==value)
db_session.query(db_model).filter(filter_args).first()
```

On looking at my code again, it would be better as:
```
def get_by(self, db_session: Session, **kwargs: Any) -> Optional[Base]:
        filter_args = [getattr(self.db_model, k)==v for k, v in kwargs.items()]
        return db_session.query(self.db_model).filter(*filter_args).first()
```

And if it's too confusing maybe it's better not to use it in this example project anyway!

Reply:
Thanks a lot for the explanation :pray: Not used to SQL alchemy syntax yet.

I think the comprehension list with the array is simpler to understand. I will go with it.

I am in the process or updating this PR with the ORM feature added lately. I will integrate your suggestions in the process 😄

Reply:
Thanks!, I appreciate the work involved in that

Reply:
I came across a simpler way, you don't need to change it or anything, just sharing a simpler solution.
```
def get_by(self, db_session: Session, **kwargs: Any) -> Optional[Base]:
        return db_session.query(self.db_model).filter_by(kwargs).first()
```

Reply:
I have added `get_first_by` and `get_multi_by` based on your suggestion 🍰